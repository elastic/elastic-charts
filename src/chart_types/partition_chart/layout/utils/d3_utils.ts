import { rgb as d3Rgb, RGBColor as D3RGBColor } from 'd3-color';

type RGB = number;
type A = number;
export type RgbTuple = [RGB, RGB, RGB, RGB?];
export type RgbObject = { r: RGB; g: RGB; b: RGB; opacity: A };

const defaultColor: RgbObject = { r: 255, g: 0, b: 0, opacity: 1 };
const defaultD3Color: D3RGBColor = d3Rgb(defaultColor.r, defaultColor.g, defaultColor.b, defaultColor.opacity);

export function stringToRGB(cssColorSpecifier: string): RgbObject {
  return d3Rgb(cssColorSpecifier) || defaultColor;
}

function argsToRGB(r: number, g: number, b: number, opacity: number): D3RGBColor {
  return d3Rgb(r, g, b, opacity) || defaultD3Color;
}

export function argsToRGBString(r: number, g: number, b: number, opacity: number): string {
  // d3.rgb returns an Rgb instance, which has a specialized `toString` method
  return argsToRGB(r, g, b, opacity).toString();
}

export function RGBtoString(rgb: RgbObject): string {
  const { r, g, b, opacity } = rgb;
  return argsToRGBString(r, g, b, opacity);
}
