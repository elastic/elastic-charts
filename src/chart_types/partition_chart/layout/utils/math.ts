import { wrapToTau } from '../geometry';

export const TAU = 2 * Math.PI;
export const RIGHT_ANGLE = TAU / 4;
export const GOLDEN_RATIO = 1.618;

export function trueBearingToStandardPositionAngle(alphaIn: number) {
  return wrapToTau(RIGHT_ANGLE - alphaIn);
}

export function logarithm(base: number, y: number) {
  return Math.log(y) / Math.log(base);
}
