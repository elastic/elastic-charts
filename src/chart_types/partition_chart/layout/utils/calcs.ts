/*
 * Licensed to Elasticsearch B.V. under one or more contributor
 * license agreements. See the NOTICE file distributed with
 * this work for additional information regarding copyright
 * ownership. Elasticsearch B.V. licenses this file to you under
 * the Apache License, Version 2.0 (the "License"); you may
 * not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License. */

import { Ratio } from '../types/geometry_types';
import { RgbTuple, stringToRGB } from './d3_utils';
import { Color } from '../../../../utils/commons';

/** @internal */
export function hueInterpolator(colors: RgbTuple[]) {
  return (d: number) => {
    const index = Math.round(d * 255);
    const [r, g, b, a] = colors[index];
    return colors[index].length === 3 ? `rgb(${r},${g},${b})` : `rgba(${r},${g},${b},${a})`;
  };
}

/** @internal */
export function addOpacity(hexColorString: string, opacity: Ratio) {
  // this is a super imperfect multiplicative alpha blender that assumes a "#rrggbb" or "#rrggbbaa" hexColorString
  // todo roll some proper utility that can handle "rgb(...)", "rgba(...)", "red", {r, g, b} etc.
  return opacity === 1
    ? hexColorString
    : hexColorString.slice(0, 7) +
        (hexColorString.slice(7).length === 0 || parseInt(hexColorString.slice(7, 2), 16) === 255
          ? `00${Math.round(opacity * 255).toString(16)}`.substr(-2) // color was of full opacity
          : `00${Math.round((parseInt(hexColorString.slice(7, 2), 16) / 255) * opacity * 255).toString(16)}`.substr(
              -2,
            ));
}

/** @internal */
export function arrayToLookup(keyFun: Function, array: Array<any>) {
  return Object.assign({}, ...array.map((d) => ({ [keyFun(d)]: d })));
}

/** @internal */
export function colorIsDark(color: Color) {
  // fixme this assumes a white or very light background
  const rgba = stringToRGB(color);
  const { r, g, b, opacity } = rgba;
  const a = rgba.hasOwnProperty('opacity') ? opacity : 1;
  return r * 0.299 + g * 0.587 + b * 0.114 < a * 150;
}

/** @internal */
export function getFillTextColor(shapeFillColor: Color, textColor: Color, textInvertible: boolean) {
  const { r: tr, g: tg, b: tb, opacity: to } = stringToRGB(textColor);
  const backgroundIsDark = colorIsDark(shapeFillColor);
  const specifiedTextColorIsDark = colorIsDark(textColor);
  const inverseForContrast = textInvertible && specifiedTextColorIsDark === backgroundIsDark;
  return inverseForContrast
    ? to === undefined
      ? `rgb(${255 - tr}, ${255 - tg}, ${255 - tb})`
      : `rgba(${255 - tr}, ${255 - tg}, ${255 - tb}, ${to})`
    : textColor;
}

/** @internal */
export function integerSnap(n: number) {
  return Math.floor(n);
}

type NumberMap = (n: number) => number;

/** @internal */
export function monotonicHillClimb(
  getResponse: NumberMap,
  maxVar: number,
  responseUpperConstraint: number,
  domainSnap: NumberMap = (n: number) => n,
  minVar: number = 0,
) {
  let loVar = domainSnap(minVar);
  const loResponse = getResponse(loVar);
  let hiVar = domainSnap(maxVar);
  let hiResponse = getResponse(hiVar);

  if (loResponse > responseUpperConstraint || loVar > hiVar) {
    // bail if even the lowest value doesn't satisfy the constraint
    return NaN;
  }

  if (hiResponse <= responseUpperConstraint) {
    return hiVar; // early bail if maxVar is compliant
  }

  let pivotVar: number = NaN;
  let pivotResponse: number = NaN;
  let lastPivotResponse: number = NaN;
  while (loVar < hiVar) {
    const newPivotVar = (loVar + hiVar) / 2;
    const newPivotResponse = getResponse(domainSnap(newPivotVar));
    if (newPivotResponse === pivotResponse || newPivotResponse === lastPivotResponse) {
      return domainSnap(loVar); // bail if we're good and not making further progress
    }
    pivotVar = newPivotVar;
    lastPivotResponse = pivotResponse; // for prevention of bistable oscillation around discretization snap
    pivotResponse = newPivotResponse;
    const pivotIsCompliant = pivotResponse <= responseUpperConstraint;
    if (pivotIsCompliant) {
      loVar = pivotVar;
    } else {
      hiVar = pivotVar;
      hiResponse = pivotResponse;
    }
  }
  return domainSnap(pivotVar);
}
