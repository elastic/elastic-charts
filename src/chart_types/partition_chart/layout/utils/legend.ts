/*
 * Licensed to Elasticsearch B.V. under one or more contributor
 * license agreements. See the NOTICE file distributed with
 * this work for additional information regarding copyright
 * ownership. Elasticsearch B.V. licenses this file to you under
 * the Apache License, Version 2.0 (the "License"); you may
 * not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

import { CategoryKey } from '../../../../common/category';
import { map } from '../../../../common/iterables';
import { LegendItem } from '../../../../common/legend';
import { LegendPositionConfig } from '../../../../specs/settings';
import { isHierarchicalLegend } from '../../../../utils/legend';
import { Layer } from '../../specs';
import { QuadViewModel } from '../types/viewmodel_types';

function makeKey(...keyParts: CategoryKey[]): string {
  return keyParts.join('---');
}

function compareTreePaths(
  { index: oiA, innerIndex: iiA, path: a }: QuadViewModel,
  { index: oiB, innerIndex: iiB, path: b }: QuadViewModel,
): number {
  if (oiA !== oiB) return oiA - oiB;
  if (iiA !== iiB) return iiA - iiB;
  for (let i = 0; i < Math.min(a.length, b.length); i++) {
    const diff = a[i].index - b[i].index;
    if (diff) {
      return diff;
    }
  }
  return a.length - b.length; // if one path is fully contained in the other, then parent (shorter) goes first
}

/** @internal */
export function getLegendItems(
  id: string,
  layers: Layer[],
  flatLegend: boolean | undefined,
  legendMaxDepth: number,
  legendPosition: LegendPositionConfig,
  quadViewModel: QuadViewModel[],
): LegendItem[] {
  const uniqueNames = new Set(map(({ dataName, fillColor }) => makeKey(dataName, fillColor), quadViewModel));
  const useHierarchicalLegend = isHierarchicalLegend(flatLegend, legendPosition);

  const formattedLabel = ({ dataName, depth }: QuadViewModel) => {
    const formatter = layers[depth - 1]?.nodeLabel;
    return formatter ? formatter(dataName) : dataName;
  };

  function compareNames(aItem: QuadViewModel, bItem: QuadViewModel): number {
    const a = formattedLabel(aItem);
    const b = formattedLabel(bItem);
    return a < b ? -1 : a > b ? 1 : 0;
  }

  const excluded: Set<string> = new Set();
  const items = quadViewModel.filter(({ depth, dataName, fillColor }) => {
    if (legendMaxDepth !== null && depth > legendMaxDepth) {
      return false;
    }
    if (!useHierarchicalLegend) {
      const key = makeKey(dataName, fillColor);
      if (uniqueNames.has(key) && excluded.has(key)) {
        return false;
      }
      excluded.add(key);
    }
    return true;
  });

  items.sort(flatLegend ? compareNames : compareTreePaths);

  return items.map<LegendItem>((item) => {
    const { dataName, fillColor, depth, path } = item;
    return {
      color: fillColor,
      label: formattedLabel(item),
      childId: dataName,
      depth: useHierarchicalLegend ? depth - 1 : 0,
      path,
      seriesIdentifiers: [{ key: dataName, specId: id }],
      keys: [],
    };
  });
}
