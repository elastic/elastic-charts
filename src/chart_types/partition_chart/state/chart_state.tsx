/*
 * Licensed to Elasticsearch B.V. under one or more contributor
 * license agreements. See the NOTICE file distributed with
 * this work for additional information regarding copyright
 * ownership. Elasticsearch B.V. licenses this file to you under
 * the Apache License, Version 2.0 (the "License"); you may
 * not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

import React, { RefObject } from 'react';

import { ChartTypes } from '../..';
import { Tooltip } from '../../../components/tooltip';
import { InternalChartState, GlobalChartState, BackwardRef } from '../../../state/chart_state';
import { InitStatus } from '../../../state/selectors/get_internal_is_intialized';
import { DebugState } from '../../../state/types';
import { Dimensions } from '../../../utils/dimensions';
import { Partition } from '../renderer/canvas/partition';
import { HighlighterFromHover } from '../renderer/dom/highlighter_hover';
import { HighlighterFromLegend } from '../renderer/dom/highlighter_legend';
import { computeLegendSelector } from './selectors/compute_legend';
import { getLegendItemsLabels } from './selectors/get_legend_items_labels';
import { isTooltipVisibleSelector } from './selectors/is_tooltip_visible';
import { createOnElementClickCaller } from './selectors/on_element_click_caller';
import { createOnElementOutCaller } from './selectors/on_element_out_caller';
import { createOnElementOverCaller } from './selectors/on_element_over_caller';
import { getPieSpec } from './selectors/pie_spec';
import { getTooltipInfoSelector } from './selectors/tooltip';

const EMPTY_MAP = new Map();

/** @internal */
export class PartitionState implements InternalChartState {
  chartType = ChartTypes.Partition;

  onElementClickCaller: (state: GlobalChartState) => void;
  onElementOverCaller: (state: GlobalChartState) => void;
  onElementOutCaller: (state: GlobalChartState) => void;

  constructor() {
    this.onElementClickCaller = createOnElementClickCaller();
    this.onElementOverCaller = createOnElementOverCaller();
    this.onElementOutCaller = createOnElementOutCaller();
  }

  isInitialized(globalState: GlobalChartState) {
    return getPieSpec(globalState) !== null ? InitStatus.Initialized : InitStatus.SpecNotInitialized;
  }

  isBrushAvailable() {
    return false;
  }

  isBrushing() {
    return false;
  }

  isChartEmpty() {
    return false;
  }

  getLegendItemsLabels(globalState: GlobalChartState) {
    return getLegendItemsLabels(globalState);
  }

  getLegendItems(globalState: GlobalChartState) {
    return computeLegendSelector(globalState);
  }

  getLegendExtraValues() {
    return EMPTY_MAP;
  }

  chartRenderer(containerRef: BackwardRef, forwardStageRef: RefObject<HTMLCanvasElement>) {
    return (
      <>
        <Tooltip getChartContainerRef={containerRef} />
        <Partition forwardStageRef={forwardStageRef} />
        <HighlighterFromHover />
        <HighlighterFromLegend />
      </>
    );
  }

  getPointerCursor() {
    return 'default';
  }

  isTooltipVisible(globalState: GlobalChartState) {
    return { visible: isTooltipVisibleSelector(globalState), isExternal: false };
  }

  getTooltipInfo(globalState: GlobalChartState) {
    return getTooltipInfoSelector(globalState);
  }

  getTooltipAnchor(state: GlobalChartState) {
    const { position } = state.interactions.pointer.current;
    return {
      isRotated: false,
      x1: position.x,
      y1: position.y,
    };
  }

  eventCallbacks(globalState: GlobalChartState) {
    this.onElementOverCaller(globalState);
    this.onElementOutCaller(globalState);
    this.onElementClickCaller(globalState);
  }

  // TODO
  getProjectionContainerArea(): Dimensions {
    return { width: 0, height: 0, top: 0, left: 0 };
  }

  // TODO
  getMainProjectionArea(): Dimensions {
    return { width: 0, height: 0, top: 0, left: 0 };
  }

  // TODO
  getBrushArea(): Dimensions | null {
    return null;
  }

  // TODO
  getDebugState(): DebugState {
    return {};
  }
}
