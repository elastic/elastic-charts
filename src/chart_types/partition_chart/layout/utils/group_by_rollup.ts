/*
 * Licensed to Elasticsearch B.V. under one or more contributor
 * license agreements. See the NOTICE file distributed with
 * this work for additional information regarding copyright
 * ownership. Elasticsearch B.V. licenses this file to you under
 * the Apache License, Version 2.0 (the "License"); you may
 * not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

import { CategoryKey } from '../../../../common/category';
import { Relation } from '../../../../common/text_utils';
import { LegendPath } from '../../../../state/actions/legend';
import { Datum, ValueAccessor } from '../../../../utils/common';

export const AGGREGATE_KEY = 'value';
export const STATISTICS_KEY = 'statistics';
export const DEPTH_KEY = 'depth';
export const CHILDREN_KEY = 'children';
export const INPUT_KEY = 'inputIndex';
export const PARENT_KEY = 'parent';
export const SORT_INDEX_KEY = 'sortIndex';
export const PATH_KEY = 'path';

export interface Statistics {
  globalAggregate: number;
}

export interface NodeDescriptor {
  [AGGREGATE_KEY]: number;
  [DEPTH_KEY]: number;
  [STATISTICS_KEY]: Statistics;
  [INPUT_KEY]?: Array<number>;
}

export type ArrayEntry = [Key, ArrayNode];
export type HierarchyOfArrays = Array<ArrayEntry>;
export interface ArrayNode extends NodeDescriptor {
  [CHILDREN_KEY]: HierarchyOfArrays;
  [PARENT_KEY]: ArrayNode;
  [SORT_INDEX_KEY]: number;
  [PATH_KEY]: LegendPath;
}

type HierarchyOfMaps = Map<Key, MapNode>;
interface MapNode extends NodeDescriptor {
  [CHILDREN_KEY]?: HierarchyOfMaps;
  [PARENT_KEY]?: ArrayNode;
}

/** @internal */
export const HIERARCHY_ROOT_KEY: Key = '__root_key__';

/** @public */
export type PrimitiveValue = string | number | null; // there could be more but sufficient for now
export type Key = CategoryKey;
export type Sorter = (a: number, b: number) => number;
type NodeSorter = (a: ArrayEntry, b: ArrayEntry) => number;

export const entryKey = ([key]: ArrayEntry) => key;
export const entryValue = ([, value]: ArrayEntry) => value;
export function depthAccessor(n: ArrayEntry) {
  return entryValue(n)[DEPTH_KEY];
}
export function aggregateAccessor(n: ArrayEntry): number {
  return entryValue(n)[AGGREGATE_KEY];
}
export function parentAccessor(n: ArrayEntry): ArrayNode {
  return entryValue(n)[PARENT_KEY];
}
export function childrenAccessor(n: ArrayEntry) {
  return entryValue(n)[CHILDREN_KEY];
}
export function sortIndexAccessor(n: ArrayEntry) {
  return entryValue(n)[SORT_INDEX_KEY];
}
export function pathAccessor(n: ArrayEntry) {
  return entryValue(n)[PATH_KEY];
}
const ascending: Sorter = (a, b) => a - b;
const descending: Sorter = (a, b) => b - a;

/** @public */
export function getNodeName(node: ArrayNode) {
  const index = node[SORT_INDEX_KEY];
  const arrayEntry: ArrayEntry = node[PARENT_KEY][CHILDREN_KEY][index];
  return entryKey(arrayEntry);
}

/** @internal */
export function groupByRollup(
  keyAccessors: Array<((a: Datum) => Key) | ((a: Datum, i: number) => Key)>,
  valueAccessor: ValueAccessor,
  {
    reducer,
    identity,
  }: {
    reducer: (prev: number, next: number) => number;
    identity: () => any;
  },
  factTable: Relation,
): HierarchyOfMaps {
  const statistics: Statistics = {
    globalAggregate: NaN,
  };
  const reductionMap: HierarchyOfMaps = factTable.reduce((p: HierarchyOfMaps, n, index) => {
    const keyCount = keyAccessors.length;
    let pointer: HierarchyOfMaps = p;
    keyAccessors.forEach((keyAccessor, i) => {
      const key: Key = keyAccessor(n, index);
      const last = i === keyCount - 1;
      const node = pointer.get(key);
      const inputIndices = node?.[INPUT_KEY] ?? [];
      const childrenMap = node?.[CHILDREN_KEY] ?? new Map();
      const aggregate = node?.[AGGREGATE_KEY] ?? identity();
      const reductionValue = reducer(aggregate, valueAccessor(n));
      pointer.set(key, {
        [AGGREGATE_KEY]: reductionValue,
        [STATISTICS_KEY]: statistics,
        [INPUT_KEY]: [...inputIndices, index],
        [DEPTH_KEY]: i,
        ...(!last && { [CHILDREN_KEY]: childrenMap }),
      });
      if (childrenMap) {
        // will always be true except when exiting from forEach, ie. upon encountering the leaf node
        pointer = childrenMap;
      }
    });
    return p;
  }, new Map());
  if (reductionMap.get(HIERARCHY_ROOT_KEY) !== undefined) {
    statistics.globalAggregate = (reductionMap.get(HIERARCHY_ROOT_KEY) as MapNode)[AGGREGATE_KEY];
  }
  return reductionMap;
}

function getRootArrayNode(): ArrayNode {
  const children: HierarchyOfArrays = [];
  const bootstrap: Omit<ArrayNode, typeof PARENT_KEY> = {
    [AGGREGATE_KEY]: NaN,
    [DEPTH_KEY]: NaN,
    [CHILDREN_KEY]: children,
    [INPUT_KEY]: [] as number[],
    [PATH_KEY]: [] as LegendPath,
    [SORT_INDEX_KEY]: 0,
    [STATISTICS_KEY]: { globalAggregate: 0 },
  };
  (bootstrap as ArrayNode)[PARENT_KEY] = bootstrap as ArrayNode;
  return bootstrap as ArrayNode; // TS doesn't yet handle bootstrapping but the `Omit` above retains guarantee for all props except `[PARENT_KEY]`
}

/** @internal */
export function mapsToArrays(root: HierarchyOfMaps, sorter: NodeSorter | null): HierarchyOfArrays {
  const groupByMap = (node: HierarchyOfMaps, parent: ArrayNode) => {
    const items = Array.from(
      node,
      ([key, value]: [Key, MapNode]): ArrayEntry => {
        const valueElement = value[CHILDREN_KEY];
        const resultNode: ArrayNode = {
          [AGGREGATE_KEY]: NaN,
          [STATISTICS_KEY]: { globalAggregate: NaN },
          [CHILDREN_KEY]: [],
          [DEPTH_KEY]: NaN,
          [SORT_INDEX_KEY]: NaN,
          [PARENT_KEY]: parent,
          [INPUT_KEY]: [],
          [PATH_KEY]: [],
        };
        const newValue: ArrayNode = Object.assign(
          resultNode,
          value,
          valueElement && { [CHILDREN_KEY]: groupByMap(valueElement, resultNode) },
        );
        return [key, newValue];
      },
    );
    if (sorter !== null) {
      items.sort(sorter);
    }
    return items.map((n: ArrayEntry, i) => {
      entryValue(n).sortIndex = i;
      return n;
    });
  }; // with the current algo, decreasing order is important
  const tree = groupByMap(root, getRootArrayNode());
  const buildPaths = ([key, mapNode]: ArrayEntry, currentPath: LegendPath) => {
    const newPath = [...currentPath, { index: mapNode[SORT_INDEX_KEY], value: key }];
    mapNode[PATH_KEY] = newPath; // in-place mutation, so disabled `no-param-reassign`
    mapNode.children.forEach((entry) => buildPaths(entry, newPath));
  };
  buildPaths(tree[0], []);
  return tree;
}

/** @internal */
export function mapEntryValue(entry: ArrayEntry) {
  return entryValue(entry)[AGGREGATE_KEY];
}

/** @internal */
export function aggregateComparator(accessor: (v: any) => any, sorter: Sorter): NodeSorter {
  return (a, b) => sorter(accessor(a), accessor(b));
}

/** @internal */
export const childOrders = {
  ascending,
  descending,
};

// type MeanReduction = { sum: number; count: number };
// type MedianReduction = Array<number>;

/** @internal */
export const aggregators = {
  one: {
    identity: () => 0,
    reducer: () => 1,
  },
  count: {
    identity: () => 0,
    reducer: (r: number) => r + 1,
  },
  sum: {
    identity: () => 0,
    reducer: (r: number, n: number) => r + n,
  },
  min: {
    identity: () => Infinity,
    reducer: (r: number, n: number) => Math.min(r, n),
  },
  max: {
    identity: () => -Infinity,
    reducer: (r: number, n: number) => Math.max(r, n),
  },
  min0: {
    identity: () => 0,
    reducer: (r: number, n: number) => Math.min(r, n),
  },
  max0: {
    identity: () => 0,
    reducer: (r: number, n: number) => Math.max(r, n),
  },
  // todo more TS typing is needed to use these
  // mean: {
  //   identity: (): MeanReduction => ({ sum: 0, count: 0 }),
  //   reducer: (r: MeanReduction, n: number) => {
  //     r.sum += n;
  //     r.count++;
  //     return r;
  //   },
  //   finalizer: (r: MeanReduction): number => r.sum / r.count,
  // },
  // median: {
  //   identity: (): MedianReduction => [],
  //   reducer: (r: MedianReduction, n: number) => {
  //     r.push(n);
  //     return r;
  //   },
  //   finalizer: (r: MedianReduction): number => {
  //     const sorted = r.sort(ascending);
  //     const len = r.length;
  //     const even = len === len % 2;
  //     const half = len / 2;
  //     return even ? (sorted[half - 1] + sorted[half]) / 2 : sorted[half - 0.5];
  //   },
  // },
};
