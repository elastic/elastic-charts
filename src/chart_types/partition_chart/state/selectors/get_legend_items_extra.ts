/*
 * Licensed to Elasticsearch B.V. under one or more contributor
 * license agreements. See the NOTICE file distributed with
 * this work for additional information regarding copyright
 * ownership. Elasticsearch B.V. licenses this file to you under
 * the Apache License, Version 2.0 (the "License"); you may
 * not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

import createCachedSelector from 're-reselect';

import { LegendItemExtraValues } from '../../../../common/legend';
import { SeriesKey } from '../../../../common/series_id';
import { SettingsSpec } from '../../../../specs';
import { getChartIdSelector } from '../../../../state/selectors/get_chart_id';
import { getSettingsSpecSelector } from '../../../../state/selectors/get_settings_specs';
import { HierarchyOfArrays, CHILDREN_KEY } from '../../layout/utils/group_by_rollup';
import { PartitionSpec } from '../../specs';
import { getPieSpec } from './pie_spec';
import { getTree } from './tree';

/** @internal */
export const getLegendItemsExtra = createCachedSelector(
  [getPieSpec, getSettingsSpecSelector, getTree],
  (pieSpec, { legendMaxDepth }, tree): Map<SeriesKey, LegendItemExtraValues> => {
    const legendExtraValues = new Map<SeriesKey, LegendItemExtraValues>();

    return pieSpec && isValidLegendMaxDepth(legendMaxDepth)
      ? getExtraValueMap(pieSpec, tree, legendMaxDepth)
      : legendExtraValues;
  },
)(getChartIdSelector);

/**
 * Check if the legendMaxDepth from settings is a valid number (NaN or <=0)
 *
 * @param legendMaxDepth - SettingsSpec['legendMaxDepth']
 */
function isValidLegendMaxDepth(legendMaxDepth: SettingsSpec['legendMaxDepth']): boolean {
  return typeof legendMaxDepth === 'number' && !Number.isNaN(legendMaxDepth) && legendMaxDepth > 0;
}

/**
 * Creates flat extra value map from nested key path
 */
function getExtraValueMap(
  { layers, valueFormatter }: Pick<PartitionSpec, 'layers' | 'valueFormatter'>,
  tree: HierarchyOfArrays,
  maxDepth: number,
  depth: number = 0,
  keys: Map<SeriesKey, LegendItemExtraValues> = new Map(),
): Map<SeriesKey, LegendItemExtraValues> {
  for (let i = 0; i < tree.length; i++) {
    const branch = tree[i];
    const [key, arrayNode] = branch;
    const { value, path, [CHILDREN_KEY]: children } = arrayNode;

    if (key != null) {
      const values: LegendItemExtraValues = new Map();
      const formattedValue = valueFormatter ? valueFormatter(value) : value;

      values.set(key, formattedValue);
      keys.set(path.map(({ index }) => index).join('__'), values);
    }

    if (depth < maxDepth) {
      getExtraValueMap({ layers, valueFormatter }, children, maxDepth, depth + 1, keys);
    }
  }
  return keys;
}
