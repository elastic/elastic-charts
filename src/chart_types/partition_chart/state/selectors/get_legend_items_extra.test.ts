/*
 * Licensed to Elasticsearch B.V. under one or more contributor
 * license agreements. See the NOTICE file distributed with
 * this work for additional information regarding copyright
 * ownership. Elasticsearch B.V. licenses this file to you under
 * the Apache License, Version 2.0 (the "License"); you may
 * not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

import { Store } from 'redux';

import { MockSeriesSpec, MockGlobalSpec } from '../../../../mocks/specs';
import { MockStore } from '../../../../mocks/store';
import { GlobalChartState } from '../../../../state/chart_state';
import { PrimitiveValue } from '../../layout/utils/group_by_rollup';
import { getLegendItemsExtra } from './get_legend_items_extra';

describe('Partition - Legend item extra values', () => {
  type TestDatum = [string, string, string, number];
  const spec = MockSeriesSpec.sunburst({
    data: [
      ['aaa', 'aa', '1', 1],
      ['aaa', 'aa', '1', 2],
      ['aaa', 'aa', '3', 1],
      ['aaa', 'bb', '4', 1],
      ['aaa', 'bb', '5', 1],
      ['aaa', 'bb', '6', 1],
      ['bbb', 'aa', '7', 1],
      ['bbb', 'aa', '8', 1],
      ['bbb', 'bb', '9', 1],
      ['bbb', 'bb', '10', 1],
      ['bbb', 'cc', '11', 1],
      ['bbb', 'cc', '12', 1],
    ],
    valueAccessor: (d: TestDatum) => d[3],
    layers: [
      {
        groupByRollup: (datum: TestDatum) => datum[0],
        nodeLabel: (d: PrimitiveValue) => String(d),
      },
      {
        groupByRollup: (datum: TestDatum) => datum[1],
        nodeLabel: (d: PrimitiveValue) => String(d),
      },
      {
        groupByRollup: (datum: TestDatum) => datum[2],
        nodeLabel: (d: PrimitiveValue) => String(d),
      },
    ],
  });
  let store: Store<GlobalChartState>;

  beforeEach(() => {
    store = MockStore.default();
  });

  it('should return all extra values in nested legend', () => {
    MockStore.addSpecs([spec], store);

    const extraValues = getLegendItemsExtra(store.getState());
    expect([...extraValues.keys()]).toEqual([
      '0',
      '0__0',
      '0__0__0',
      '0__0__0__0',
      '0__0__0__1',
      '0__0__1',
      '0__0__1__0',
      '0__0__1__1',
      '0__0__1__2',
      '0__1',
      '0__1__0',
      '0__1__0__0',
      '0__1__0__1',
      '0__1__1',
      '0__1__1__0',
      '0__1__1__1',
      '0__1__2',
      '0__1__2__0',
      '0__1__2__1',
    ]);
    expect(extraValues.values()).toMatchSnapshot();
  });

  it('should return extra values in nested legend within max depth of 1', () => {
    const settings = MockGlobalSpec.settings({ legendMaxDepth: 1 });
    MockStore.addSpecs([settings, spec], store);

    const extraValues = getLegendItemsExtra(store.getState());
    expect([...extraValues.keys()]).toEqual(['0', '0__0', '0__1']);
    expect(extraValues.values()).toMatchSnapshot();
  });

  it('should return extra values in nested legend within max depth of 2', () => {
    const settings = MockGlobalSpec.settings({ legendMaxDepth: 2 });
    MockStore.addSpecs([settings, spec], store);

    const extraValues = getLegendItemsExtra(store.getState());
    expect([...extraValues.keys()]).toEqual([
      '0',
      '0__0',
      '0__0__0',
      '0__0__1',
      '0__1',
      '0__1__0',
      '0__1__1',
      '0__1__2',
    ]);
    expect(extraValues.values()).toMatchSnapshot();
  });

  it('filters all extraValues is depth is 0', () => {
    const settings = MockGlobalSpec.settings({ legendMaxDepth: 0 });
    MockStore.addSpecs([settings, spec], store);

    const extraValues = getLegendItemsExtra(store.getState());
    expect([...extraValues.keys()]).toEqual([]);
  });

  it('filters all extraValues is depth is NaN', () => {
    const settings = MockGlobalSpec.settings({ legendMaxDepth: NaN });
    MockStore.addSpecs([settings, spec], store);

    const extraValues = getLegendItemsExtra(store.getState());
    expect([...extraValues.keys()]).toEqual([]);
  });
});
