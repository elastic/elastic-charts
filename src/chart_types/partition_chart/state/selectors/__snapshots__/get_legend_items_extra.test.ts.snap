// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Partition - Legend item extra values should return all extra values in nested legend 1`] = `Object {}`;

exports[`Partition - Legend item extra values should return extra values in nested legend within max depth of 1 1`] = `Object {}`;

exports[`Partition - Legend item extra values should return extra values in nested legend within max depth of 2 1`] = `Object {}`;
